Issue with drawing arrows in matplotlib 3.5.0

Currently when trying to use the CLI you get an error when it tries to draw the arrows:

```
[cvl62853@pc0110 scanspec]$ pipenv run scanspec plot 'Snake(Line("z", 5, 6, 2) * Line("y", 2, 3, 2) * Line("x", 1, 2, 2))'
Exception in Tkinter callback
Traceback (most recent call last):
  File "/dls_sw/prod/tools/RHEL7-x86_64/Python3/3-7-2dls1/prefix/lib/python3.7/tkinter/__init__.py", line 1705, in __call__
    return self.func(*args)
  File "/dls_sw/prod/tools/RHEL7-x86_64/Python3/3-7-2dls1/prefix/lib/python3.7/tkinter/__init__.py", line 749, in callit
    func(*args)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/backends/_backend_tk.py", line 251, in idle_draw
    self.draw()
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/backends/backend_tkagg.py", line 9, in draw
    super().draw()
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 436, in draw
    self.figure.draw(self.renderer)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/artist.py", line 73, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/artist.py", line 50, in draw_wrapper
    return draw(artist, renderer)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/figure.py", line 2804, in draw
    renderer, self, artists, self.suppressComposite)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/image.py", line 132, in _draw_list_compositing_images
    a.draw(renderer)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/matplotlib/artist.py", line 50, in draw_wrapper
    return draw(artist, renderer)
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 449, in draw
    reverse=True):
  File "/scratch/cvl62853/pipenv/scanspec-ajHh2QfZ/lib/python3.7/site-packages/mpl_toolkits/mplot3d/axes3d.py", line 436, in do_3d_projection
    return artist.do_3d_projection(renderer)
AttributeError: '_Arrow3D' object has no attribute 'do_3d_projection'
```
Issue is here: https://github.com/matplotlib/matplotlib/issues/21688

Fix was to re-add a (deprecated) method to their Arrow3D class:
```
    def do_3d_projection(self, renderer=None):
        xs3d, ys3d, zs3d = self._verts3d
        xs, ys, zs = proj3d.proj_transform(xs3d, ys3d, zs3d, self.axes.M)
        self.set_positions((xs[0], ys[0]), (xs[1], ys[1]))

        return np.min(zs)
```
This fixes the plotting, but you still get test failures for 3d plots:
```
Traceback (most recent call last):
  File "/dls/science/users/cvl62853/support/python/scanspec/tests/test_cli.py", line 212, in test_plot_3D_line
    assert len(artists) == 4
AssertionError: assert 0 == 4
  +0
  -4
```

--- a/src/scanspec/core.py
+++ b/src/scanspec/core.py
@@ -270,7 +270,7 @@ class Frames(Generic[Axis]):
         # All axespoints arrays are same length, pick the first one
         return len(self.gap)
 
-    def extract(self, indices: np.ndarray, calculate_gap=True) -> "Frames[Axis]":
+    def extract(self, indices: np.ndarray, calculate_gap=True) -> Frames[Axis]:
         """Return a new Frames object restricted to the indices provided.
 
         Args:
@@ -296,7 +296,7 @@ class Frames(Generic[Axis]):
 
         return _merge_frames(self, dict_merge=extract_dict, gap_merge=extract_gap)
 
-    def concat(self, other: "Frames[Axis]", gap: bool = False) -> "Frames[Axis]":
+    def concat(self, other: Frames[Axis], gap: bool = False) -> Frames[Axis]:
         """Return a new Frames object concatenating self and other.
 
         Requires both Frames objects to have the same axes.
@@ -327,7 +327,7 @@ class Frames(Generic[Axis]):
 
         return _merge_frames(self, other, dict_merge=concat_dict, gap_merge=concat_gap)
 
-    def zip(self, other: "Frames[Axis]") -> "Frames[Axis]":
+    def zip(self, other: Frames[Axis]) -> Frames[Axis]:
         """Return a new Frames object merging self and other.
 
         Require both Frames objects to not share axes.
@@ -392,7 +392,7 @@ class SnakedFrames(Frames[Axis]):
         self.gap[0] = False
 
     @classmethod
-    def from_frames(cls, frames: Frames[Axis]) -> "SnakedFrames[Axis]":
+    def from_frames(cls, frames: Frames[Axis]) -> SnakedFrames[Axis]:
         """Create a snaked version of a `Frames` object."""
         return cls(frames.midpoints, frames.lower, frames.upper, frames.gap)
 

--- a/src/scanspec/plot.py
+++ b/src/scanspec/plot.py
@@ -35,6 +35,14 @@ class _Arrow3D(patches.FancyArrowPatch):
         self.set_positions((xs[0], ys[0]), (xs[1], ys[1]))
         super().draw(renderer)
 
+    # Added here because of https://github.com/matplotlib/matplotlib/issues/21688
+    def do_3d_projection(self, renderer=None):
+        xs3d, ys3d, zs3d = self._verts3d
+        xs, ys, zs = proj3d.proj_transform(xs3d, ys3d, zs3d, self.axes.M)
+        self.set_positions((xs[0], ys[0]), (xs[1], ys[1]))
+
+        return np.min(zs)
+
 
 def _plot_arrow(axes, arrays: List[np.ndarray]):
     if len(arrays) == 1:

--- a/src/scanspec/regions.py
+++ b/src/scanspec/regions.py
@@ -47,16 +47,16 @@ class Region(Generic[Axis]):
         """Produce a mask of which points are in the region."""
         raise NotImplementedError(self)
 
-    def __or__(self, other) -> "UnionOf[Axis]":
+    def __or__(self, other) -> UnionOf[Axis]:
         return if_instance_do(other, Region, lambda o: UnionOf(self, o))
 
-    def __and__(self, other) -> "IntersectionOf[Axis]":
+    def __and__(self, other) -> IntersectionOf[Axis]:
         return if_instance_do(other, Region, lambda o: IntersectionOf(self, o))
 
-    def __sub__(self, other) -> "DifferenceOf[Axis]":
+    def __sub__(self, other) -> DifferenceOf[Axis]:
         return if_instance_do(other, Region, lambda o: DifferenceOf(self, o))
 
-    def __xor__(self, other) -> "SymmetricDifferenceOf[Axis]":
+    def __xor__(self, other) -> SymmetricDifferenceOf[Axis]:
         return if_instance_do(other, Region, lambda o: SymmetricDifferenceOf(self, o))
 
 

--- a/src/scanspec/specs.py
+++ b/src/scanspec/specs.py
@@ -1,3 +1,5 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
 from typing import Any, Callable, Dict, Generic, List, Mapping, Optional
 
@@ -77,23 +79,23 @@ class Spec(Generic[Axis]):
         """Return `Midpoints` that can be iterated point by point."""
         return Midpoints(self.calculate(bounds=False))
 
-    def __rmul__(self, other) -> "Product[Axis]":
+    def __rmul__(self, other) -> Product[Axis]:
         return if_instance_do(other, int, lambda o: Product(Repeat(o), self))
 
-    def __mul__(self, other) -> "Product[Axis]":
+    def __mul__(self, other) -> Product[Axis]:
         return if_instance_do(other, Spec, lambda o: Product(self, o))
 
-    def __and__(self, other) -> "Mask[Axis]":
+    def __and__(self, other) -> Mask[Axis]:
         return if_instance_do(other, Region, lambda o: Mask(self, o))
 
-    def __invert__(self) -> "Snake[Axis]":
+    def __invert__(self) -> Snake[Axis]:
         return Snake(self)
 
-    def zip(self, other: "Spec") -> "Zip[Axis]":
+    def zip(self, other: Spec) -> Zip[Axis]:
         """`Zip` the Spec with another, iterating in tandem."""
         return Zip(self, other)
 
-    def concat(self, other: "Spec") -> "Concat[Axis]":
+    def concat(self, other: Spec) -> Concat[Axis]:
         """`Concat` the Spec with another, iterating one after the other."""
         return Concat(self, other)
 
@@ -102,7 +104,7 @@ class Spec(Generic[Axis]):
         return serialize(Spec, self)
 
     @classmethod
-    def deserialize(cls, serialized: Mapping[str, Any]) -> "Spec[Axis]":
+    def deserialize(cls, serialized: Mapping[str, Any]) -> Spec[Axis]:
         """Deserialize the spec from a dictionary."""
         return deserialize(cls, serialized)
 
@@ -308,18 +310,18 @@ class Mask(Spec[Axis]):
 
     # *+ bind more tightly than &|^ so without these overrides we
     # would need to add brackets to all combinations of Regions
-    def __or__(self, other: "Region[Axis]") -> "Mask[Axis]":
+    def __or__(self, other: Region[Axis]) -> Mask[Axis]:
         return if_instance_do(other, Region, lambda o: Mask(self.spec, self.region | o))
 
-    def __and__(self, other: "Region[Axis]") -> "Mask[Axis]":
+    def __and__(self, other: Region[Axis]) -> Mask[Axis]:
         return if_instance_do(other, Region, lambda o: Mask(self.spec, self.region & o))
 
-    def __xor__(self, other: "Region[Axis]") -> "Mask[Axis]":
+    def __xor__(self, other: Region[Axis]) -> Mask[Axis]:
         return if_instance_do(other, Region, lambda o: Mask(self.spec, self.region ^ o))
 
     # This is here for completeness, tends not to be called as - binds
     # tighter than &
-    def __sub__(self, other: "Region[Axis]") -> "Mask[Axis]":
+    def __sub__(self, other: Region[Axis]) -> Mask[Axis]:
         return if_instance_do(other, Region, lambda o: Mask(self.spec, self.region - o))
 
 
@@ -500,7 +502,7 @@ class Line(Spec[Axis]):
             float, schema(description="Upper bound of the last point of the line")
         ],
         num: ANum,
-    ) -> "Line[Axis]":
+    ) -> Line[Axis]:
         """Specify a Line by extreme bounds instead of midpoints.
 
         .. example_spec::
@@ -541,7 +543,7 @@ class Static(Spec[Axis]):
     def duration(
         duration: A[float, schema(description="The duration of each static point")],
         num: ANum = 1,
-    ) -> "Static[str]":
+    ) -> Static[str]:
         """A static spec with no motion, only a duration repeated "num" times.
 
         .. example_spec::
@@ -626,7 +628,7 @@ class Spiral(Spec[Axis]):
         rotate: A[
             float, schema(description="How much to rotate the angle of the spiral"),
         ] = 0.0,
-    ) -> "Spiral[Axis]":
+    ) -> Spiral[Axis]:
         """Specify a Spiral equally spaced in "x_axis" and "y_axis".
 
         .. example_spec::

--- a/src/scanspec/sphinxext.py
+++ b/src/scanspec/sphinxext.py
@@ -14,7 +14,7 @@ def always_create_figures():
     """
     orig_f = plot_directive.out_of_date
     # Patch the plot directive so it thinks all sources are out of date
-    plot_directive.out_of_date = lambda o, d: True
+    plot_directive.out_of_date = lambda *args, **kwargs: True
     try:
         yield
     finally:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -205,13 +205,6 @@ def test_plot_3D_line() -> None:
     assert_min_max_3d(lines[11], 1, 2, 2, 3, 5, 6, length=8)
     # End
     assert_min_max_3d(lines[12], 0.5, 0.5, 2, 2, 6, 6)
-    # Arrows
-    artists = axes.artists
-    assert len(artists) == 4
-    assert_3d_arrow(artists[0], 0.5, 2, 5)
-    assert_3d_arrow(artists[1], 2.5, 3, 5)
-    assert_3d_arrow(artists[2], 0.5, 3, 6)
-    assert_3d_arrow(artists[3], 2.5, 2, 6)
 
 
 def test_schema() -> None:

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -7,6 +7,7 @@ import pytest
 from click.testing import CliRunner
 
 from scanspec import cli
+from scanspec.plot import _Arrow3D
 
 
 def assert_min_max_2d(line, xmin, xmax, ymin, ymax, length=None):
@@ -205,6 +206,16 @@ def test_plot_3D_line() -> None:
     assert_min_max_3d(lines[11], 1, 2, 2, 3, 5, 6, length=8)
     # End
     assert_min_max_3d(lines[12], 0.5, 0.5, 2, 2, 6, 6)
+    # Arrows
+    extra_artists = axes.get_default_bbox_extra_artists()
+    arrow_artists = list(
+        filter(lambda artist: isinstance(artist, _Arrow3D), extra_artists)
+    )
+    assert len(arrow_artists) == 4
+    assert_3d_arrow(arrow_artists[0], 0.5, 2, 5)
+    assert_3d_arrow(arrow_artists[1], 2.5, 3, 5)
+    assert_3d_arrow(arrow_artists[2], 0.5, 3, 6)
+    assert_3d_arrow(arrow_artists[3], 2.5, 2, 6)
 
 
 def test_schema() -> None:

